enterprise/dbcrypt/dbcrypt.go

```diff
@@ -0,0 +1,296 @@
+//
+// The Cipher is currently used to encrypt/decrypt the following fields:
+// - database.UserLink.OAuthAccessToken
+// - database.UserLink.OAuthRefreshToken
+// - database.GitAuthLink.OAuthAccessToken
+// - database.GitAuthLink.OAuthRefreshToken
+// - database.DBCryptSentinelValue
+//
+// Encrypted fields are stored in the following format:
+// "dbcrypt-${b64encode(<first 7 digits of cipher's SHA256 digest>-<encrypted value>)}"
+//
+// The first 7 characters of the cipher's SHA256 digest are used to identify the cipher
+// used to encrypt the value.
+//
+// Multiple ciphers can be provided to support key rotation. The primary cipher is used
+// to encrypt and decrypt all data. Secondary ciphers are only used for decryption.
+// We currently only use a single secondary cipher.
+package dbcrypt
+
+import (
+	"context"
+	"database/sql"
+	"encoding/base64"
+	"errors"
+	"strings"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+)
+
+// MagicPrefix is prepended to all encrypted values in the database.
+// This is used to determine if a value is encrypted or not.
+// If it is encrypted but a key is not provided, an error is returned.
+// MagicPrefix will be followed by the first 7 characters of the cipher's
+// SHA256 digest, followed by a dash, followed by the base64-encoded
+// encrypted value.
+const MagicPrefix = "dbcrypt-"
+
+// sentinelValue is the value that is stored in the database to indicate
+// whether encryption is enabled. If not enabled, the value either not
+// present, or is the raw string "coder".
+// Otherwise, the value must be the encrypted value of the string "coder"
+// using the current cipher.
+const sentinelValue = "coder"
+
+var (
+	ErrNotEnabled       = xerrors.New("encryption is not enabled")
+	ErrSentinelMismatch = xerrors.New("database is already encrypted under a different key")
+	b64encode           = base64.StdEncoding.EncodeToString
+	b64decode           = base64.StdEncoding.DecodeString
+)
+
+// DecryptFailedError is returned when decryption fails.
+// It unwraps to sql.ErrNoRows.
+type DecryptFailedError struct {
+	Inner error
+}
+
+func (e *DecryptFailedError) Error() string {
+	return xerrors.Errorf("decrypt failed: %w", e.Inner).Error()
+}
+
+func (*DecryptFailedError) Unwrap() error {
+	return sql.ErrNoRows
+}
+
+// New creates a database.Store wrapper that encrypts/decrypts values
+// stored at rest in the database.
+func New(ctx context.Context, db database.Store, cs *Ciphers) (database.Store, error) {
+	if cs == nil {
+		return nil, xerrors.Errorf("no ciphers configured")
+	}
+	dbc := &dbCrypt{
+		ciphers: cs,
+		Store:   db,
+	}
+	// nolint: gocritic // This is allowed.
+	if err := ensureEncrypted(dbauthz.AsSystemRestricted(ctx), dbc); err != nil {
+		return nil, xerrors.Errorf("ensure encrypted database fields: %w", err)
+	}
+	return dbc, nil
+}
+
+type dbCrypt struct {
+	ciphers *Ciphers
```

Comment:
Does this need to care about whether it's ciphers or just a single cipher?

Reply:
If someone accidentally plumbs in an `*AES256` by accident it would result in a regression. I'm tempted to unexport the single cipher and only export `Ciphers` instead.